From 55e1d07dbbb58c6d7088c16dd6394a2fa13e172e Mon Sep 17 00:00:00 2001
From: Bryan Sandoz <17123735+skyclmbr@users.noreply.github.com>
Date: Thu, 30 Apr 2026 21:43:11 -0700
Subject: [PATCH 01/23] feat(vr): Phase 0 VR scaffold (VSP_VR, vsp_vr, --vr
 stub)

Add optional VSP_VR build, static vsp_vr stubs, and guarded --vr entry that prints a placeholder until OpenXR runtime is wired.

Made-with: Cursor

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 48a03c316..95cc3356a 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -14,6 +14,9 @@ MESSAGE( STATUS "CMake version: ${CMAKE_VERSION}" )
 
 SET( CMAKE_CXX_STANDARD 17 )
 
+# Optional VR scaffolding. Runtime features are added incrementally.
+option( VSP_VR "Build with experimental VR scaffolding" OFF )
+
 IF(CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64" OR CMAKE_SYSTEM_PROCESSOR MATCHES "amd64")
   SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
   SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
@@ -162,6 +165,10 @@ SET( VSP_G_LIBRARIES_CORE_FIRST
     ${VSP_COMMON_LIBRARIES}
 )
 
+IF( VSP_VR AND NOT VSP_NO_GRAPHICS )
+	ADD_SUBDIRECTORY( vr )
+ENDIF()
+
 ADD_SUBDIRECTORY( vsp )
 
 set(IS_NUMPY_FOUND FALSE)
diff --git a/src/vr/CMakeLists.txt b/src/vr/CMakeLists.txt
new file mode 100644
index 000000000..44b0ac5b7
--- /dev/null
+++ b/src/vr/CMakeLists.txt
@@ -0,0 +1,7 @@
+add_library( vsp_vr STATIC
+    VRInterfaces.cpp
+    VRMode.cpp
+)
+
+target_compile_definitions( vsp_vr PUBLIC VSP_VR_ENABLED )
+target_include_directories( vsp_vr PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} )
diff --git a/src/vr/VRInterfaces.cpp b/src/vr/VRInterfaces.cpp
new file mode 100644
index 000000000..9063e539d
--- /dev/null
+++ b/src/vr/VRInterfaces.cpp
@@ -0,0 +1,69 @@
+//
+// This file is released under the terms of the NASA Open Source Agreement (NOSA)
+// version 1.3 as detailed in the LICENSE file which accompanies this software.
+//
+
+#include "VRInterfaces.h"
+
+VRManager &VRManager::GetInstance()
+{
+    static VRManager inst;
+    return inst;
+}
+
+void VRManager::Init() {}
+void VRManager::Shutdown() {}
+
+VRRenderer &VRRenderer::GetInstance()
+{
+    static VRRenderer inst;
+    return inst;
+}
+
+void VRRenderer::Init() {}
+void VRRenderer::Shutdown() {}
+
+VRControllerInput &VRControllerInput::GetInstance()
+{
+    static VRControllerInput inst;
+    return inst;
+}
+
+void VRControllerInput::Init() {}
+void VRControllerInput::Shutdown() {}
+
+VRInteraction &VRInteraction::GetInstance()
+{
+    static VRInteraction inst;
+    return inst;
+}
+
+void VRInteraction::Init() {}
+void VRInteraction::Shutdown() {}
+
+VRRoomScale &VRRoomScale::GetInstance()
+{
+    static VRRoomScale inst;
+    return inst;
+}
+
+void VRRoomScale::Init() {}
+void VRRoomScale::Shutdown() {}
+
+VRPassthrough &VRPassthrough::GetInstance()
+{
+    static VRPassthrough inst;
+    return inst;
+}
+
+void VRPassthrough::Init() {}
+void VRPassthrough::Shutdown() {}
+
+VRPanelManager &VRPanelManager::GetInstance()
+{
+    static VRPanelManager inst;
+    return inst;
+}
+
+void VRPanelManager::Init() {}
+void VRPanelManager::Shutdown() {}
diff --git a/src/vr/VRInterfaces.h b/src/vr/VRInterfaces.h
new file mode 100644
index 000000000..54e30fcf9
--- /dev/null
+++ b/src/vr/VRInterfaces.h
@@ -0,0 +1,81 @@
+//
+// This file is released under the terms of the NASA Open Source Agreement (NOSA)
+// version 1.3 as detailed in the LICENSE file which accompanies this software.
+//
+
+#ifndef VRINTERFACES_H
+#define VRINTERFACES_H
+
+// Phase 0: stub interfaces; OpenXR and runtime wiring come in later phases.
+
+class VRManager
+{
+public:
+    static VRManager &GetInstance();
+    void Init();
+    void Shutdown();
+private:
+    VRManager() = default;
+};
+
+class VRRenderer
+{
+public:
+    static VRRenderer &GetInstance();
+    void Init();
+    void Shutdown();
+private:
+    VRRenderer() = default;
+};
+
+class VRControllerInput
+{
+public:
+    static VRControllerInput &GetInstance();
+    void Init();
+    void Shutdown();
+private:
+    VRControllerInput() = default;
+};
+
+class VRInteraction
+{
+public:
+    static VRInteraction &GetInstance();
+    void Init();
+    void Shutdown();
+private:
+    VRInteraction() = default;
+};
+
+class VRRoomScale
+{
+public:
+    static VRRoomScale &GetInstance();
+    void Init();
+    void Shutdown();
+private:
+    VRRoomScale() = default;
+};
+
+class VRPassthrough
+{
+public:
+    static VRPassthrough &GetInstance();
+    void Init();
+    void Shutdown();
+private:
+    VRPassthrough() = default;
+};
+
+class VRPanelManager
+{
+public:
+    static VRPanelManager &GetInstance();
+    void Init();
+    void Shutdown();
+private:
+    VRPanelManager() = default;
+};
+
+#endif
diff --git a/src/vr/VRMode.cpp b/src/vr/VRMode.cpp
new file mode 100644
index 000000000..8e171be4f
--- /dev/null
+++ b/src/vr/VRMode.cpp
@@ -0,0 +1,16 @@
+//
+// This file is released under the terms of the NASA Open Source Agreement (NOSA)
+// version 1.3 as detailed in the LICENSE file which accompanies this software.
+//
+
+#include "VRMode.h"
+
+#include <cstdio>
+
+int RunVRMode( int argc, char **argv )
+{
+    ( void )argc;
+    ( void )argv;
+    printf( "[VSP_VR] VR mode scaffold is enabled, but runtime is not implemented yet.\n" );
+    return 1;
+}
diff --git a/src/vr/VRMode.h b/src/vr/VRMode.h
new file mode 100644
index 000000000..ef8190d02
--- /dev/null
+++ b/src/vr/VRMode.h
@@ -0,0 +1,11 @@
+//
+// This file is released under the terms of the NASA Open Source Agreement (NOSA)
+// version 1.3 as detailed in the LICENSE file which accompanies this software.
+//
+
+#ifndef VRMODE_H
+#define VRMODE_H
+
+int RunVRMode( int argc, char **argv );
+
+#endif
diff --git a/src/vsp/CMakeLists.txt b/src/vsp/CMakeLists.txt
index 1cad6c9de..114ce4c12 100644
--- a/src/vsp/CMakeLists.txt
+++ b/src/vsp/CMakeLists.txt
@@ -42,6 +42,10 @@ IF( NOT VSP_NO_GRAPHICS )
         ${VSP_G_LIBRARIES_CORE_FIRST}
     )
 
+    IF( VSP_VR )
+        TARGET_LINK_LIBRARIES( vsp vsp_vr )
+    ENDIF()
+
     INSTALL( TARGETS vsp RUNTIME DESTINATION . )
 
     ADD_EXECUTABLE(apitest_g
diff --git a/src/vsp/main.cpp b/src/vsp/main.cpp
index 2489b09a2..3a73c8805 100644
--- a/src/vsp/main.cpp
+++ b/src/vsp/main.cpp
@@ -7,6 +7,11 @@
 //#include "vld.h"
 
 #include <cstdio>
+#include <cstring>
+
+#ifdef VSP_VR_ENABLED
+#include "VRMode.h"
+#endif
 
 #ifdef WIN32
 #include <windows.h>
@@ -200,6 +205,22 @@ void ThreadCheckVersionNumber()
 
 int main( int argc, char** argv )
 {
+#ifdef VSP_VR_ENABLED
+    bool vr_mode = false;
+    for ( int i = 1; i < argc; ++i )
+    {
+        if ( argv[i] && strcmp( argv[i], "--vr" ) == 0 )
+        {
+            vr_mode = true;
+            break;
+        }
+    }
+    if ( vr_mode )
+    {
+        return RunVRMode( argc, argv );
+    }
+#endif
+
     // Set up MainThreadID if this is entry point.
     MainThreadIDMgr.getInstance();
 
-- 
2.54.0.windows.1

